### ISSUE TITLE
enhancement: need more error-handling options

### ISSUE THREAD

**Issue by boneskull** [vstirbu/fsm-as-promised#14]
given setup:

```
const fsm = StateMachine({
  events: [
    {name: 'foo', from: 'none', to: 'done'}
  ],
  final: 'done'
});
```

Let's say `foo()` is part of your public API.  The user calls `fsm.foo()` and everything's fine; but then the user calls `fsm.foo()` again, resulting in an "invalid state transition"-type Error.  AFAIK, there's no way to customize how this is handled.

The naive solution here would be adding more supported callbacks.

**Comment by vstirbu** [comment]
Given the example, the behaviour is appropriate. `foo` is a transition between states and one started there cannot be other simultaneous transitions.

You can check if you can do a transition before triggering using `can` and `cannot` properties of the fsm (I noticed that they are not documented in README).

**Comment by boneskull** [comment]
OK, that should suffice.  thanks

**Comment by boneskull** [comment]
actually, wait, no, that would require the user to check manually.

**Comment by boneskull** [comment]
I agree that yes, the second call to `fsm.foo()` should result in an error.  however, I need to:

- change the error messaging to hide implementation details (the user doesn't care if this is a state machine)
- and run some arbitrary code when this does happen.

otherwise I'm going to have to wrap the FSM and use it behind-the-scenes, which I was hoping to avoid.

**Comment by vstirbu** [comment]
Can you share a specific example of what you try to achieve?

The library keeps track in a naive way if it is [inTransition](https://github.com/vstirbu/fsm-as-promised/blob/master/lib/index.js#L31). That can be reimplemented in a more sophisticated manner using a stack like structure so that you could do something more when handling errors than what is possible right now.

**Comment by boneskull** [comment]
Roughly:

- a `Test` is an FSM.  It begins in state `idle`
- when user calls `Test#run()`, the FSM transitions to the `running` state
- in the `running` state, the next state transition can be one of `failed`, `skipped`, `errored` or `passed`
- if the `Test` reaches `passed` or `errored`, these are final states
- if the `Test` reaches `skipped` or `failed`, the user can transition into the `running` state again via `Test#run()`
- If a user erroneously calls `Test#run()`--or writes code which erroneously calls it, more likely--on a `Test` in one of the two final states, I would like to give the user more contextual information about why this is not possible.
- In addition, I would like to record the failed *attempt* to transition from say, `passed` into `running` in an internal data structure

**Comment by vstirbu** [comment]
Wouldn't this behaviour be easier to handle if the `fsm` would be an event emitter? You could implement all that logic externally while keeping the implementation minimal.

Alternatively we could add one more _optional_ callback [here](https://github.com/vstirbu/fsm-as-promised/blob/master/lib/index.js#L301).

**Comment by boneskull** [comment]
Which logic are you referring to?

**Comment by boneskull** [comment]
FSM-as-EE wouldn't allow me to catch a transition error before it reaches my user.  It may address the last point I made above, but not much else.

I think I wouldn't even have created this issue if I had more control over `StateMachine`'s behavior.  As it stands, everything's in a closure, which is problematic for hackability.

hmm, I have an idea... will follow up.

**Comment by yarandoo** [comment]
Shall this issue be closed with the merging of #51?
